File: touch.go
package main

import (
	"unsafe"

	. "github.com/linux-surface/iptsd/processing"
	. "github.com/linux-surface/iptsd/protocol"
)

func IptsTouchHandleHeatmap(ipts *IptsContext, heatmap *Heatmap) error {
	touch := ipts.Devices.Touch
	points := touch.Processor.Inputs(heatmap)
	blocked := false

	if ipts.Config.BlockOnPalm {
		for i := 0; i < len(points); i++ {
			blocked = blocked || points[i].IsPalm
		}
	}

	for i := 0; i < len(points); i++ {
		p := points[i]

		touch.Device.Emit(EV_ABS, ABS_MT_SLOT, int32(p.Slot))

		if p.Index != -1 && !p.IsStable {
			continue
		}

		if p.IsPalm || blocked {
			touch.Device.Emit(EV_ABS, ABS_MT_TRACKING_ID, -1)
			touch.Device.Emit(EV_ABS, ABS_MT_POSITION_X, 0)
			touch.Device.Emit(EV_ABS, ABS_MT_POSITION_Y, 0)

			continue
		}

		touch.Device.Emit(EV_ABS, ABS_MT_TRACKING_ID, int32(p.Index))
		touch.Device.Emit(EV_ABS, ABS_MT_POSITION_X, int32(p.X))
		touch.Device.Emit(EV_ABS, ABS_MT_POSITION_Y, int32(p.Y))
	}

	err := touch.Device.Emit(EV_SYN, SYN_REPORT, 0)
	if err != nil {
		return err
	}

	return nil
}

func IptsTouchHandleInput(ipts *IptsContext, frame IptsPayloadFrame) error {
	size := uint32(0)
	var hm *Heatmap

	for size < frame.Size {
		report, err := ipts.Protocol.ReadReport()
		if err != nil {
			return err
		}

		size += uint32(report.Size) + uint32(unsafe.Sizeof(report))

		switch report.Type {
		case IPTS_REPORT_TYPE_TOUCH_HEATMAP_DIM:
			height, err := ipts.Protocol.ReadByte()
			if err != nil {
				break
			}

			width, err := ipts.Protocol.ReadByte()
			if err != nil {
				break
			}

			dev := ipts.Devices.Touch
			hm = dev.Processor.GetHeatmap(int(width), int(height))

			err = ipts.Protocol.Skip(6)
			break
		case IPTS_REPORT_TYPE_TOUCH_HEATMAP:
			err = ipts.Protocol.Read(hm.Data)
			break
		default:
			// ignored
			err = ipts.Protocol.Skip(uint32(report.Size))
			break
		}

		if err != nil {
			return nil
		}
	}

	if hm == nil {
		return nil
	}

	err := IptsTouchHandleHeatmap(ipts, hm)
	if err != nil {
		return err
	}

	return nil
}

File: touch.c
// SPDX-License-Identifier: GPL-2.0-or-later

#include <linux/uinput.h>
#include <stdint.h>
#include <string.h>

#include "context.h"
#include "devices.h"
#include "heatmap.h"
#include "protocol.h"
#include "touch.h"
#include "touch-processing.h"
#include "utils.h"

static void iptsd_touch_emit_empty(int dev)
{
	iptsd_devices_emit(dev, EV_ABS, ABS_MT_TRACKING_ID, -1);
	iptsd_devices_emit(dev, EV_ABS, ABS_MT_POSITION_X, 0);
	iptsd_devices_emit(dev, EV_ABS, ABS_MT_POSITION_Y, 0);
}

static void iptsd_touch_emit(int dev, struct iptsd_touch_input in)
{
	iptsd_devices_emit(dev, EV_ABS, ABS_MT_TRACKING_ID, in.index);
	iptsd_devices_emit(dev, EV_ABS, ABS_MT_POSITION_X, in.x);
	iptsd_devices_emit(dev, EV_ABS, ABS_MT_POSITION_Y, in.y);
}

static int iptsd_touch_handle_heatmap(struct iptsd_context *iptsd,
		struct heatmap *hm)
{
	struct iptsd_touch_device touch = iptsd->devices.touch;
	bool blocked = false;

	iptsd_touch_processing_inputs(&touch.processor, hm);

	if (iptsd->config.block_on_palm) {
		for (int i = 0; i < touch.processor.max_contacts; i++)
			blocked = blocked || touch.processor.inputs[i].is_palm;
	}

	for (int i = 0; i < touch.processor.max_contacts; i++) {
		struct iptsd_touch_input in = touch.processor.inputs[i];

		iptsd_devices_emit(touch.dev, EV_ABS, ABS_MT_SLOT, in.slot);
		if (in.index != -1 && !in.is_stable)
			continue;

		if (in.is_palm || blocked) {
			iptsd_touch_emit_empty(touch.dev);
			continue;
		}

		iptsd_touch_emit(touch.dev, in);
	}

	int ret = iptsd_devices_emit(touch.dev, EV_SYN, SYN_REPORT, 0);
	if (ret < 0)
		iptsd_err(ret, "Failed to emit touch report");

	return ret;
}

int iptsd_touch_handle_input(struct iptsd_context *iptsd,
		struct ipts_payload_frame *frame)
{
	int pos = 0;
	struct heatmap *hm = NULL;

	while (pos < frame->size) {
		struct ipts_report *report =
			(struct ipts_report *)&frame->data[pos];

		uint8_t width = 0;
		uint8_t height = 0;

		switch (report->type) {
		case IPTS_REPORT_TYPE_TOUCH_HEATMAP_DIM:
			height = report->data[0];
			width = report->data[1];

			hm = iptsd_touch_processing_get_heatmap(
					&iptsd->devices.touch.processor,
					width, height);
			break;
		case IPTS_REPORT_TYPE_TOUCH_HEATMAP:
			memcpy(hm->data, report->data, hm->size);
			break;
		}

		pos += report->size + sizeof(struct ipts_report);
	}

	if (!hm)
		return 0;

	int ret = iptsd_touch_handle_heatmap(iptsd, hm);
	if (ret < 0)
		iptsd_err(ret, "Failed to handle touch heatmap");

	return ret;
}

